ND Filter calibration product

juliamilton opened the pull request:
## Describe your changes

**nd_filter_calibration.py: ND Filter Calibration Product Generation**
The ND filter calibration product is created by using observations of dim standard stars observed without the ND filter and bright standard stars observed with the ND filter in place. The core idea is to measure a flux calibration factor from the dim stars, for which the expected flux is known from reference spectra (specific stars used are defined in fluxcal.py), and then use that factor to compute the optical density of the ND filter from measurements of bright stars taken through that same filter.

1. Flux Calibration Factor (No ND Filter): Dim stars, with well-characterized flux levels, are observed without the ND filter. In each image, the script computes a centroid of the star and sums the total flux within a small aperture around that centroid. This measured flux (in electrons per second) is compared to the known/ expected flux from reference data to derive a calibration factor C in erg s−1 cm−2/(e− s−1)\mathrm{erg\,s^{-1}\,cm^{-2}} / (\mathrm{e^{-}\,s^{-1}})ergs−1cm−2/(e−s−1). Background is also subtracted in this step.
2. Optical Density Computation (With ND Filter): Next, bright stars are observed through the ND filter. By using the same aperture-based flux measurement procedure and applying the calibration factor C, the script converts measured flux (electrons per second) to physical flux units. Dividing this measured flux by the expected flux (from the same reference star databases) yields the ND filter OD. This OD indicates the attenuation provided by the filter.
3. Saving Results: The script then groups results by target star, checks variation in OD to see if it meets a specified uniformity threshold, and writes each bright star’s OD values and centroid positions (x, y) into a new FITS file saved as an NDFilterCalibration data type. These FITS products are the calibration outputs needed to document the performance of the ND filter at different star positions.

**test_nd_filter_calibration.py: ND Filter Calibration Product Test Script**
The test script automates the generation of mock datasets and then calls the main ND filter calibration script.
1. Mock Data Creation:
    * Dim Star Datasets: The script simulates ten “dim” standard stars by creating synthetic Gaussian sources with relatively faint flux. Each simulated image is saved with metadata indicating the star name (from the fluxcal.py star catalog), filter name, exposure time, etc. These serve as the baseline for computing the calibration factor without ND filter.
    * Bright Star Datasets: The script then simulates four bright standard stars in a 3×3 raster pattern each (for a total of nine images per star). These images use a higher flux to approximate observations with an ND filter in place.
2. Header & FITS Packaging: Every mock image includes relevant headers (e.g., TARGET, CFAMNAME for filter name, EXPTIME, etc.) and is saved to a FITS file as a NDFilterCalibration datatype. 

**data.py: New NDFilterCalibration datatype**
A new NDFilterCalibration datatype is defined in data.py.

## Type of change

Please delete options that are not relevant (and this line).

- New feature (non-breaking change which adds functionality)


## Reference any relevant issues (don't forget the #)
Produce ND filter calibration #242


## Checklist before requesting a review
- [X ] I have linted my code
- [X] I have verified that all unit tests pass in a clean environment and added new unit tests, as needed
- [X] I have verified that all docstrings are properly formatted and added new documentation, as needed

JuergenSchreiber commented:
@juliamilton Hi Julia, nice work. There are overlaps of course with the flux calibration, I see that you already use some functions from there, and this might get even more. I am working on the flux cal factors calibration product, and I am thinking about whether we should generate one for each filter band or put in a list of all available filter bands in only one product. I think you NDFilterCalibration product is only valid for one band, or am I wrong? This is also a question to the other reviewers how to organize this.

semaphoreP commented:
I would assume that the calibration does need to be done on a per-band basis. Note that we don't have so many bands, but I think it is evitable we need to do this on a per-band basis. 

I agree that it would be great for Julia and Juergen to sync up, since your efforts are linked (as you note).

semaphoreP commented:
@juliamilton can you fix the docstrings, to get the tests running?

juliamilton commented:
@semaphoreP I have no idea why it is failing the linter. The doc strings all seem properly formatted to me and it passes with a 0 output when I run it on my computer. Would you be able to take a quick look and let me know if you can point me to any info why it fails here?

semaphoreP commented:
I think you need to replace "Parameters" with "Args" and that should fix nearly all if not all the errors (see other functions, you need the exact format)

juliamilton commented:
@semaphoreP thanks! It's now failing a few e2e tests because it says "OBSID keyword not found". This makes sense, because we removed that keyword from the L1 input files and replaced it with OBSNUM, and I'd guess the github tests are using a new version of L1 files. Many of these tests (like bp_map_e2e) pass on my computer with the old input files, and others (like flatfield_e2e) fail due to issues specific to the code that currently exists in those functions, which I did not write and are not part of this PR. How would you like me to handle testing/ debugging my code given this?

semaphoreP commented:
Do we need to change OBSID -> OBSNUM for this PR? Can that be a separate PR that does that change and fixes the tests that fail because of it?

Note that the e2e tests aren't even being run. It's the importing of modules that seem to be breaking due to this change.

juliamilton commented:
Hi @JuergenSchreiber, thanks for your comments and sorry for the delay. It seems that future implementation of this ND filter calibration code will ultimately change quite a bit, but for now I will address your comments. 

- Thanks for the note about the calculate_band_irradiance function in fluxcal.py. I have gotten rid of my redundant one now.
- I think your method of doing the aperture and flux cal factor makes sense logically, but it appears to use many FITS keywords that we do plan to include in the L1 files (CRPIX1, CRPIX2, CTYPE1, CTYPE2, CDELT1, CDELT2, CRVAL1, CRVAL2). We can discuss this more, perhaps on slack or in a future meeting. For now, I will continue to use the method in my code since it does not rely on these keywords. I believe it is true, however, that when we get to this stage of the pipeline background should be subtracted already so that part of my code is not useful.
- Yes, the way my code is written will require a separate calibration file for each filter band. I have updated the calibration file name to include the filter used.

juliamilton commented:
@semaphoreP The issue was that it was creating a default entry with OBSNUM instead of OBSID in caldb since, unlike when you run it locally, there is nothing in the caldb already. I reverted that. Beyond that though, it's not required to change OBSID -> OBSNUM in the rest of the code for this PR. Is there an existing ticket to align the keywords in the DRP code with those currently in the L1 headers? If not, I could start working on a PR for that.

juliamilton commented:
@semaphoreP @maxwellmb All automated checks pass now, I've addressed Juergen's comments (with the caveat that this particular calibration code will likely change significantly in the next development cycle), and the unit tests are documented here: https://collaboration.ipac.caltech.edu/display/romancoronagraph/DRP+Test+Definition . Unless there are are further comments, I think it can be merged to dev.

JuergenSchreiber commented:
> Hi @JuergenSchreiber, thanks for your comments and sorry for the delay. It seems that future implementation of this ND filter calibration code will ultimately change quite a bit, but for now I will address your comments.
> 
> * Thanks for the note about the calculate_band_irradiance function in fluxcal.py. I have gotten rid of my redundant one now.

You use compute_centroid, there is a centroid function in astrom.py, maybe you could replace, too?

> * I think your method of doing the aperture and flux cal factor makes sense logically, but it appears to use many FITS keywords that we do plan to include in the L1 files (CRPIX1, CRPIX2, CTYPE1, CTYPE2, CDELT1, CDELT2, CRVAL1, CRVAL2). We can discuss this more, perhaps on slack or in a future meeting. For now, I will continue to use the method in my code since it does not rely on these keywords. I believe it is true, however, that when we get to this stage of the pipeline background should be subtracted already so that part of my code is not useful.

We have already an astrometric calibration (astrom.py)  I guess in l2b_to_l3.py, so we should have a WCS header already, since flux cal is planned in l4_to_tda.py.
It seems that we do not have a stand alone background subtraction step, so we should indeed subtract the background of an annulus during flux extraction, see my #294.

> * Yes, the way my code is written will require a separate calibration file for each filter band. I have updated the calibration file name to include the filter used.
I used the color filter name and the ND name, there seem to be 2 ND filters ND250 and ND475 in FPAM and in FSAM. I assumed both ND filters could be set somehow but I am not sure, does somebody know how this will be handled?

semaphoreP commented:
> @semaphoreP The issue was that it was creating a default entry with OBSNUM instead of OBSID in caldb since, unlike when you run it locally, there is nothing in the caldb already. I reverted that. Beyond that though, it's not required to change OBSID -> OBSNUM in the rest of the code for this PR. Is there an existing ticket to align the keywords in the DRP code with those currently in the L1 headers? If not, I could start working on a PR for that.

I think this falls under issue #140. Since you've already mostly figured out the issue, totally feel free to do a PR for it (hopefully it's short!).

semaphoreP commented:
> > * I think your method of doing the aperture and flux cal factor makes sense logically, but it appears to use many FITS keywords that we do plan to include in the L1 files (CRPIX1, CRPIX2, CTYPE1, CTYPE2, CDELT1, CDELT2, CRVAL1, CRVAL2). We can discuss this more, perhaps on slack or in a future meeting. For now, I will continue to use the method in my code since it does not rely on these keywords. I believe it is true, however, that when we get to this stage of the pipeline background should be subtracted already so that part of my code is not useful.
> 
> We have already an astrometric calibration (astrom.py) I guess in l2b_to_l3.py, so we should have a WCS header already, since flux cal is planned in l4_to_tda.py. According to Jason, it seems that we do not have a stand alone background subtraction step, so we should indeed subtract the background of an annulus during flux extraction, see my #294.


While it is true we have it, if it's not really needed to use the astrometric calibration to extract photometry, then it's better not to use it. It's good to reduce inter-dependency when it's not necessary (in case we have a problem in the astrometric calibration file).

juliamilton commented:
@JuergenSchreiber Thanks, I updated my code so it uses the astrom.py centroiding function.

I'll keep in the background subtraction step in my code for now, when your background subtraction function has been merged I will use it in my code as well for consistency.

In terms of whether the FPAM or FSAM ND filter will be set - the requirement for this calibration product specifies that it is for the FPAM. I think in the conops it can be specified that the FSAM ND filter should not be in place as well prior to acquiring the images.

@semaphoreP - could you clarify your comment? Are you suggesting I continue to use my own method of background subtraction and flux measurement in this code rather than use the methods that are/ will be developed in Juergen's files?

maxwellmb reviewed (changes requested):
Hi Julia, 

This is an excellent start on this, thanks for all your efforts! I've put in a few comments. They're mostly bigger-picture/structural, but hopefully don't require making too many small-scale changes. I think in the end it makes sense for us to use the same aperture photometry functions for this and fluxcal. Finally, the tests need to have a few more defined tests, more than just that. it doesn't crash. I recommend reviewing some of the other test files to get a good idea of what we need here. I'm happy to work through some of these changes with you, so please feel free to reach out if you want some extra input. 

Cheers, 
Max

maxwellmb commented:
Potentially helpful: I just merged in PR #294, Juergen's update with an optional local background subtraction for aperture photometry.

juliamilton commented on corgidrp/data.py:
```diff
@@ -1261,6 +1261,183 @@ def __init__(self,data_or_filepath, pri_hdr=None,ext_hdr=None, input_dataset=Non
+            raise ValueError("File that was loaded is not labeled as an NDFilterCalibration file.")
+
+
+class FluxcalFactor(Image):
```
I think it was the result of a merge conflict. I removed one instance now.

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,333 @@
+    return x_center, y_center
+
+
+def compute_flux_in_image(image_data, x_center, y_center, radius=5,
```
There was some discussion about this above, but the aper_phot function as written uses header keywords (CRPIX1, CRPIX2, CTYPE1, CTYPE2, CDELT1, CDELT2, CRVAL1, CRVAL2) that we don't currently plan to have in the L1 files. I am unclear on whether Juergen and I should work together to modify aper_phot given this information and your input about x,y center above, or whether these keywords will be added in to the headers at a higher level of processing and the flux cal and nd filter cal should be using those higher level files and not L1 files. Could you provide input?

maxwellmb replied:
I'm not opposed to using those header keywords. They're standard fits keywords, so I think that's fine. However, adding optional inputs of x and y center positions would likely sidestep this issue for now, right?

juliamilton replied:
@JuergenSchreiber's method of aper_phot requires these keywords to place the aperture. Alex is currently not planning to provide these keywords in the L1 headers so I don't think we should rely on methods that use them. Using x and y center gets around this, yes. I can update the aper_phot code to use x and y center to place the aperture if everyone is on board (I'm not sure what implications this will have for Juergen's upstream or downstream code, if any).

maxwellmb replied:
Here's my suggestion: you make a kwarg called centrer_method or something like that that allows us to choose between "WCS" or "centroiding", or "user" (or something like that). I think your centroiding is the best default as it requires the least number of assumptions, previous calculations or user inputs. 

As a general note, we can definitely rely on keywords that we expect to have in the headers even if they're not there in the L1 headers. For example, the WCS header keywords you're talking about are likely to be added by the WCS step function.

juliamilton replied:
Ok, that's fine, I can do that method you suggest! Although I think I'm fundamentally confused about what level data product inputs calibration products should be using. (Ie - L1, L2, L3, L4? Other?) How do we find out what steps have been executed prior to our calibration product running and what the input should look like in terms of the processing that has already been done?

maxwellmb replied:
well if we need to pop a step function into a recipe to make it work, then I don't see why we couldn't add WCS information. However in this case I think defaulting to no WCS is probably the way to go. For a given step function I think the philosophy should probably be to complete it with as little outside information as needed, but also that we can make it happen if we need to have something run in an earlier step function. Most of the requirements define things that need to be done, not things that can't be done if we see fit.

juliamilton replied:
I've now changed both photometry functions to accept kwargs that allow you choose between a WCS method and a centroid method to place the aperture. Centroid method is the default.

semaphoreP commented:
> @JuergenSchreiber Thanks, I updated my code so it uses the astrom.py centroiding function.
> 
> I'll keep in the background subtraction step in my code for now, when your background subtraction function has been merged I will use it in my code as well for consistency.
> 
> In terms of whether the FPAM or FSAM ND filter will be set - the requirement for this calibration product specifies that it is for the FPAM. I think in the conops it can be specified that the FSAM ND filter should not be in place as well prior to acquiring the images.
> 
> @semaphoreP - could you clarify your comment? Are you suggesting I continue to use my own method of background subtraction and flux measurement in this code rather than use the methods that are/ will be developed in Juergen's files?

I think I'm suggesting it would be nice to use flux extraction code that doesn't depend on the astrometric calibration, so there's one less thing we're depending on blindly (when the information is already in the image without needing it). But there seems to be a lot of reviewers already, so I'll let others who have thought about this more have the final say. I don't have a strong opinion... just suggesting a possible nice-to-have.

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,333 @@
+    return net_flux
+
+
+def compute_expected_flux(star_name, filter_name):
```
Yes I could, but I'd need to change it slightly and then I wouldn't use that exact function in my code. The calibrate_fluxcal_aper and calibrate_fluxcal_gauss2d functions currently use calculate_band_flux and my function uses calculate_band_irradiance (I've renamed my function compute_expected_band_irradiance for clarity). I used band integrated irradiance for my application because I'm interested in total energy over the entire filter band so I can compare expected vs actual photon counts in a straightforward way.

maxwellmb replied:
So it seems like there's only one line of code difference? Otherwise it seems like both steps share a common flow: `get_filter_name`, `read_filter_curve`, `get_calspec_file`, `read_cal_spec`. If it's just the one line difference at the end, we could have a common function that does the first bit and then the fluxcal and your code could just have that last line different, right? Possibly in the main body of the code?

juliamilton replied:
I've rewritten this function so that it uses only functions from fluxcal.py and is just a helper function for my own code.

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,333 @@
+                max_serial = current_serial
+
+    # Save calibration products
+    for star_name, star_data in flux_results.items():
```
I agree with you, but I think we need more clarification in the conops on how this calibration product should be developed. (It also sounds like the conops may change significantly, but I wanted to get some initial code written and submitted for this sprint.)

Currently, we perform a raster for each bright target. Since the targets have different magnitudes and fluxes, the OD is calculated separately for each star using the ratio of measured flux / expected flux * calibration factor. We then calculate an average OD over the raster pattern and flag it if the OD values at individual sample points deviate too much. If that happens, I’d suggest specifying in the conops that we sample a different point on the filter and re-take the calibration.

Once we obtain an average OD for each bright target, should we then average those to get a single OD value for the band and ND filter? If not, how else would you propose deriving a single OD value?

maxwellmb replied:
Perhaps I'm confused! I thought the goal of this was to create a spatial map of the OD since we expect that there might be spatial variations in it. If one location is different than other locations, then we keep track of it, but I don't think I understand why that would trigger a new calibration. 

Either way, I'm pretty sure we want the behavior of this function to still be the same: for any number of input frames (multiple stars or not within a single dataset), we still only want a single calibration file as the output. We definitely can reject specific measurements if we think they're spurious.

juliamilton replied:
Ok, I've put all bright star raster locations in a single calibration file now. I have a flag that gets set if the OD's at different locations vary by more than some threshold value; we can decide in the future what, if anything, gets done with that information.

On the other item: we will not create a spatial OD map of the whole filter. Per the requirements, we just raster around in a 3x3 grid in the area of a single point on the ND filter using the FSM. At each FSM raster move, the excam x and y coordinates of the spot and the calculated OD there are saved in a "sweet spot" dataset. From the discussions we've had, I think this procedure is to account for FPAM nonrepeatability by sampling in a small-ish local location, rather than trying to make a large map of the ND filter.

juliamilton commented on tests/test_nd_filter_calibration.py:
```diff
@@ -0,0 +1,250 @@
+]
+
+
+def create_flux_image(flux, fwhm, background, nx=1024, ny=1024):
```
I could and will in the future, but the current create_flux_image has a lot of hard-coded properties that I don't want for my images, and also generates header keywords that we don't plan on having in the L1 files (per another comment in this thread). Juergen currently uses those header keywords for his aper_phot function, so I think we first need to resolve how aper_phot will be done and whether we need to request (or derive somehow) these header keywords.

maxwellmb replied:
Having near-duplicate functions in mutliple places is a recipe for disaster. I think making some of the hardcoded values into keyword arguments is fine and you can tweak the function to accommodate both your needs. At its base level your two functions are basically doing the same thing. As discussed elsewhere having those keywords isn't a big deal as we expect them to be added in L3-L4.

juliamilton replied:
Ok, I modified the create_flux_image in mocks.py slightly and I use that now

maxwellmb commented:
> > @JuergenSchreiber Thanks, I updated my code so it uses the astrom.py centroiding function.
> > I'll keep in the background subtraction step in my code for now, when your background subtraction function has been merged I will use it in my code as well for consistency.
> > In terms of whether the FPAM or FSAM ND filter will be set - the requirement for this calibration product specifies that it is for the FPAM. I think in the conops it can be specified that the FSAM ND filter should not be in place as well prior to acquiring the images.
> > @semaphoreP - could you clarify your comment? Are you suggesting I continue to use my own method of background subtraction and flux measurement in this code rather than use the methods that are/ will be developed in Juergen's files?
> 
> I think I'm suggesting it would be nice to use flux extraction code that doesn't depend on the astrometric calibration, so there's one less thing we're depending on blindly (when the information is already in the image without needing it). But there seems to be a lot of reviewers already, so I'll let others who have thought about this more have the final say. I don't have a strong opinion... just suggesting a possible nice-to-have.

In the PR I was suggesting that we merge the two aperture photometry functions to allow for either automatic finding with header info or have an x-y input. I suppose the third option could be have the aperture photometry itself do the centroiding, as @juliamilton does it now. That might be the best default option to be honest.

JuergenSchreiber commented:
@juliamilton, please tweak the fluxcal code to your needs, also the mocks flux image generation. This was kept very simple. You might then also need to tweak some fluxcal tests.
I like the idea to have different options to find the star position. Although a completely free position finding might get problems when you have several sources in the field or some cosmics. 
Addressing the difference with calculate_band_flux or calculate_band_irradiance: calculate_band_flux calculates a 'mean' flux in calspec units at a "mean" reference wavelength within the band (see Gordon et al. (2022)). calculate_band_irradiance integrates over the band and changes the unit. In my opinion and experience (but I am not really an expert in this and you should also ask others) there is no real difference in this, especially if you consider relative fluxes. Why would you need photon counts?

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,333 @@
+    return expected_flux
+
+
+def group_by_target(dataset_entries):
```
I've rewritten this so that it uses the Dataset.split_dataset() function and is just a helper function for my own code.

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,333 @@
+    return grouped_files
+
+
+def compute_flux_calibration_factor(dim_stars_paths):
```
I use Juergen's method now and my function simply averages all the flux calibration factors for the different dim stars used as inputs.

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,333 @@
+    cal_factor = compute_flux_calibration_factor(dim_stars_paths)
+    
+    # Step 2: Group bright star files
+    grouped_files = group_by_target(bright_stars_paths)
```
As written, this calibration product is only meant to be done with one filter at a time. Different observations will need to be done with different filters to make additional calibration products specific to those filters. I am grouping by target because I need to know that target's light curve from its calspec files in order to calculate its expected flux in the bandpass.

juliamilton commented on tests/test_nd_filter_calibration.py:
```diff
@@ -0,0 +1,250 @@
+    return image
+
+
+def save_image_to_fits(image, filename):
```
Removed

juliamilton commented on tests/test_nd_filter_calibration.py:
```diff
@@ -0,0 +1,250 @@
+    return file_paths
+
+
+def mock_bright_dataset_files(output_path, bright_exptime, filter, OD):
```
I put in the option to save them to disk now or to not do that (sometimes I just like to see them when testing)

juliamilton commented on tests/test_nd_filter_calibration.py:
```diff
@@ -0,0 +1,250 @@
+    return file_paths
+
+
+if __name__ == "__main__":
```
ok, saved them as test_<names> now

JuergenSchreiber commented:
@juliamilton Hi Julia, I have a question concerning the functions calibrate_... where you calculate the fluxcal factor: What is the reason that you multiply the flux with the exposure time?  The L3 data is already normalized by the exposure time, see issue #3 and function divide_by_exptime in l2b_to_l3.py. So I was actually assuming that we start aperture photometry with units photoelectrons/s.

juliamilton commented:
> @juliamilton Hi Julia, I have a question concerning the functions calibrate_... where you calculate the fluxcal factor: What is the reason that you multiply the flux with the exposure time? The L3 data is already normalized by the exposure time, see issue #3 and function divide_by_exptime in l2b_to_l3.py. So I was actually assuming that we start aperture photometry with units photoelectrons/s. Do you try to solve the conflicts or should I also look into?

Hi @JuergenSchreiber thanks for looking into this. I had asked about which level data product should be used as input for this calibration function, but I’m still unsure about the answer. In other calibration workflows I’ve worked with, L1 data was typically used as input, so I’m not certain what processing steps - if any - will have already been applied by the time these functions are run. Could you share any documentation or reasoning that led you to conclude that we should be using L3 data as input here? I’d appreciate any clarification you can provide.

I'll look at the merge conflicts today but it would be great if you can also let me know if you see anything that you think should not be changed as I'm suggesting or if you have other feedback.

JuergenSchreiber commented:
> > @juliamilton Hi Julia, I have a question concerning the functions calibrate_... where you calculate the fluxcal factor: What is the reason that you multiply the flux with the exposure time? The L3 data is already normalized by the exposure time, see issue #3 and function divide_by_exptime in l2b_to_l3.py. So I was actually assuming that we start aperture photometry with units photoelectrons/s. Do you try to solve the conflicts or should I also look into?
> 
> Hi @JuergenSchreiber thanks for looking into this. I had asked about which level data product should be used as input for this calibration function, but I’m still unsure about the answer. In other calibration workflows I’ve worked with, L1 data was typically used as input, so I’m not certain what processing steps - if any - will have already been applied by the time these functions are run. Could you share any documentation or reasoning that led you to conclude that we should be using L3 data as input here? I’d appreciate any clarification you can provide.
> 
> I'll look at the merge conflicts today but it would be great if you can also let me know if you see anything that you think should not be changed as I'm suggesting or if you have other feedback.

Great, thanks, did you forget the background parameter in the mock function?
I see it as follows but Max should also comment: photometry is always one of the last steps in a pipeline, so it should rely on a lot of calibrations done before, and time normalization is an easy one and available, we can just add it to the walker recipe.

maxwellmb commented:
Hi Julia,

Yes you should assume that any pre-calibration steps that are needed have
already been run (assuming a step function exists). In fact, it would be
really helpful to add which steps you're assuming have been run to the
docstring. When we go to make the end to end tests we'll create recipes
that run all the steps first and have this step last, as Juergens
mentioned.

Cheers,
Max

On Thu, Feb 13, 2025, 06:46 JuergenSchreiber ***@***.***>
wrote:

> @juliamilton <https://github.com/juliamilton> Hi Julia, I have a question
> concerning the functions calibrate_... where you calculate the fluxcal
> factor: What is the reason that you multiply the flux with the exposure
> time? The L3 data is already normalized by the exposure time, see issue #3
> <https://github.com/roman-corgi/corgidrp/issues/3> and function
> divide_by_exptime in l2b_to_l3.py. So I was actually assuming that we start
> aperture photometry with units photoelectrons/s. Do you try to solve the
> conflicts or should I also look into?
>
> Hi @JuergenSchreiber <https://github.com/JuergenSchreiber> thanks for
> looking into this. I had asked about which level data product should be
> used as input for this calibration function, but I’m still unsure about the
> answer. In other calibration workflows I’ve worked with, L1 data was
> typically used as input, so I’m not certain what processing steps - if any
> - will have already been applied by the time these functions are run. Could
> you share any documentation or reasoning that led you to conclude that we
> should be using L3 data as input here? I’d appreciate any clarification you
> can provide.
>
> I'll look at the merge conflicts today but it would be great if you can
> also let me know if you see anything that you think should not be changed
> as I'm suggesting or if you have other feedback.
>
> Great, thanks, did you forget the background parameter in the mock
> function?
> I see it as follows but Max should also comment: photometry is always one
> of the last steps in a pipeline, so it should rely on a lot of calibrations
> done before, and time normalization is an easy one and available, we can
> just add it to the walker recipe.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/roman-corgi/corgidrp/pull/279#issuecomment-2656823824>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEK2SJFQCEIBV2F37AIHS2T2PSV6HAVCNFSM6AAAAABUSSIG7KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNJWHAZDGOBSGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: JuergenSchreiber]*JuergenSchreiber* left a comment
> (roman-corgi/corgidrp#279)
> <https://github.com/roman-corgi/corgidrp/pull/279#issuecomment-2656823824>
>
> @juliamilton <https://github.com/juliamilton> Hi Julia, I have a question
> concerning the functions calibrate_... where you calculate the fluxcal
> factor: What is the reason that you multiply the flux with the exposure
> time? The L3 data is already normalized by the exposure time, see issue #3
> <https://github.com/roman-corgi/corgidrp/issues/3> and function
> divide_by_exptime in l2b_to_l3.py. So I was actually assuming that we start
> aperture photometry with units photoelectrons/s. Do you try to solve the
> conflicts or should I also look into?
>
> Hi @JuergenSchreiber <https://github.com/JuergenSchreiber> thanks for
> looking into this. I had asked about which level data product should be
> used as input for this calibration function, but I’m still unsure about the
> answer. In other calibration workflows I’ve worked with, L1 data was
> typically used as input, so I’m not certain what processing steps - if any
> - will have already been applied by the time these functions are run. Could
> you share any documentation or reasoning that led you to conclude that we
> should be using L3 data as input here? I’d appreciate any clarification you
> can provide.
>
> I'll look at the merge conflicts today but it would be great if you can
> also let me know if you see anything that you think should not be changed
> as I'm suggesting or if you have other feedback.
>
> Great, thanks, did you forget the background parameter in the mock
> function?
> I see it as follows but Max should also comment: photometry is always one
> of the last steps in a pipeline, so it should rely on a lot of calibrations
> done before, and time normalization is an easy one and available, we can
> just add it to the walker recipe.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/roman-corgi/corgidrp/pull/279#issuecomment-2656823824>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEK2SJFQCEIBV2F37AIHS2T2PSV6HAVCNFSM6AAAAABUSSIG7KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNJWHAZDGOBSGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

juliamilton commented:
> Hi Julia, Yes you should assume that any pre-calibration steps that are needed have already been run (assuming a step function exists). In fact, it would be really helpful to add which steps you're assuming have been run to the docstring. When we go to make the end to end tests we'll create recipes that run all the steps first and have this step last, as Juergens mentioned. Cheers, Max
> […](#)
> On Thu, Feb 13, 2025, 06:46 JuergenSchreiber ***@***.***> wrote: @juliamilton <https://github.com/juliamilton> Hi Julia, I have a question concerning the functions calibrate_... where you calculate the fluxcal factor: What is the reason that you multiply the flux with the exposure time? The L3 data is already normalized by the exposure time, see issue #3 <#3> and function divide_by_exptime in l2b_to_l3.py. So I was actually assuming that we start aperture photometry with units photoelectrons/s. Do you try to solve the conflicts or should I also look into? Hi @JuergenSchreiber <https://github.com/JuergenSchreiber> thanks for looking into this. I had asked about which level data product should be used as input for this calibration function, but I’m still unsure about the answer. In other calibration workflows I’ve worked with, L1 data was typically used as input, so I’m not certain what processing steps - if any - will have already been applied by the time these functions are run. Could you share any documentation or reasoning that led you to conclude that we should be using L3 data as input here? I’d appreciate any clarification you can provide. I'll look at the merge conflicts today but it would be great if you can also let me know if you see anything that you think should not be changed as I'm suggesting or if you have other feedback. Great, thanks, did you forget the background parameter in the mock function? I see it as follows but Max should also comment: photometry is always one of the last steps in a pipeline, so it should rely on a lot of calibrations done before, and time normalization is an easy one and available, we can just add it to the walker recipe. — Reply to this email directly, view it on GitHub <[#279 (comment)](https://github.com/roman-corgi/corgidrp/pull/279#issuecomment-2656823824)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AEK2SJFQCEIBV2F37AIHS2T2PSV6HAVCNFSM6AAAAABUSSIG7KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNJWHAZDGOBSGQ> . You are receiving this because you were mentioned.Message ID: ***@***.***> [image: JuergenSchreiber]*JuergenSchreiber* left a comment ([roman-corgi/corgidrp#279](https://github.com/roman-corgi/corgidrp/pull/279)) <[#279 (comment)](https://github.com/roman-corgi/corgidrp/pull/279#issuecomment-2656823824)> @juliamilton <https://github.com/juliamilton> Hi Julia, I have a question concerning the functions calibrate_... where you calculate the fluxcal factor: What is the reason that you multiply the flux with the exposure time? The L3 data is already normalized by the exposure time, see issue #3 <#3> and function divide_by_exptime in l2b_to_l3.py. So I was actually assuming that we start aperture photometry with units photoelectrons/s. Do you try to solve the conflicts or should I also look into? Hi @JuergenSchreiber <https://github.com/JuergenSchreiber> thanks for looking into this. I had asked about which level data product should be used as input for this calibration function, but I’m still unsure about the answer. In other calibration workflows I’ve worked with, L1 data was typically used as input, so I’m not certain what processing steps - if any - will have already been applied by the time these functions are run. Could you share any documentation or reasoning that led you to conclude that we should be using L3 data as input here? I’d appreciate any clarification you can provide. I'll look at the merge conflicts today but it would be great if you can also let me know if you see anything that you think should not be changed as I'm suggesting or if you have other feedback. Great, thanks, did you forget the background parameter in the mock function? I see it as follows but Max should also comment: photometry is always one of the last steps in a pipeline, so it should rely on a lot of calibrations done before, and time normalization is an easy one and available, we can just add it to the walker recipe. — Reply to this email directly, view it on GitHub <[#279 (comment)](https://github.com/roman-corgi/corgidrp/pull/279#issuecomment-2656823824)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AEK2SJFQCEIBV2F37AIHS2T2PSV6HAVCNFSM6AAAAABUSSIG7KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNJWHAZDGOBSGQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Thanks @maxwellmb, I appreciate the clarification, but I’m still confused about the logic here and I may just be missing some necessary background information. I originally wrote everything assuming that the input files to this calibration function would be L1 frames, as that aligns with how other calibration products in the pipeline are documented. Given that, I wasn’t expecting any "necessary pre-calibration steps" to have already been applied.

Could you clarify why we should assume the inputs to this calibration function are not L1 files? Is there documentation specifying that we should be using L3 files instead, as Juergen is suggesting? If so, where can I find details on what processing steps will have already been applied to L3 files before this stage? I want to make sure I fully understand the expected state of the input data before proceeding.

maxwellmb commented:
It's more a matter of how the DRP is structured, with an lean towards modularization.The main processing steps are strung together with recipes rather than multiple data processing steps happening in a given step function. 

Exactly what needs to be done for each calibration product very much depends on that calibration product and it's not always fully defined ahead of time. If it's not fully defined that also gives us some freedom to make choices ourselves about what needs to be done. A quick look at the L2B->L3 steps suggests to me that we should be at least L3 for this. Happy to chat more if you want to go into details together.

juliamilton commented:
@maxwellmb Thanks for the explanation. I think I’ll need more clarification on the code base structure/ philosophy here for my future reference—for example, if most processing is supposed to be modularized and occur outside of these calibration product step functions, we may want to take out things like background subtraction from being an integrated part of these steps. But that discussion can wait.

For now, could you clarify which L2B → L3 steps you’re referring to? From what I see, the only function in that part of the code is the exposure time normalization. Based on this, I’ll remove the divide-by-exposure-time step in @JuergenSchreiber’s functions and update my code accordingly, assuming that the input files to the ND filter calibration already have exposure time normalization applied. I’ll also document any other assumptions about data format I use.

JuergenSchreiber commented:
Just trying to clarify even more: I investigated whether there is a background subtraction step planned in the pipeline, but this was not confirmed, and it is not described in the requirements as far as I know, so we better do it here. Maybe you could have a look at the e2e tests to see how the calibration generation generally works, essentially the pipeline is run up to  a certain level (+ needed other pipeline steps) and then the calibration file is generated using these products. For the ND calibration the flatfielding step is very essential, I guess.

maxwellmb commented:
> @maxwellmb Thanks for the explanation. I think I’ll need more clarification on the code base structure/ philosophy here for my future reference—for example, if most processing is supposed to be modularized and occur outside of these calibration product step functions, we may want to take out things like background subtraction from being an integrated part of these steps. But that discussion can wait.
> 

A background subtraction step function would not be an unreasonable thing to have! Let's not do it now, but it sounds like you're ve got the idea. 

> For now, could you clarify which L2B → L3 steps you’re referring to? From what I see, the only function in that part of the code is the exposure time normalization. Based on this, I’ll remove the divide-by-exposure-time step in @JuergenSchreiber’s functions and update my code accordingly, assuming that the input files to the ND filter calibration already have exposure time normalization applied. I’ll also document any other assumptions about data format I use.

Yeah just the exposure time and WCS steps in L2b to L3 is what I was thinking.

juliamilton commented:
Hi @maxwellmb and @semaphoreP, could you re-review this code? I believe I have addressed all feedback. The updated code does:

1. Calibration with Dim Stars: Calculates an average calibration factor from dim star images using aperture or PSF photometry to compare measured photo-electrons with expected band-irradiance.

2. Processing Bright Stars: Bright images are grouped by target star so OD can be calculated using expected flux for that specific star. For each target, the code verifies metadata, calculates the star’s centroid, performs photometry, and computes the optical density (OD) using the formula OD = –log₁₀(measured flux / expected flux). The resulting [OD, x, y] values form a “sweet spot” dataset.

3. Aggregating Data: All target-specific sweet spot arrays are combined into a single master Nx3 dataset, and an overall average OD and standard deviation are calculated. If the standard deviation is above a threshold, the dataset is flagged.

4. Using the Dataset to Optionally Calculate the OD of an Additional Clean Frame: If provided, a clean frame’s centroid is adjusted using a transformation matrix, and its OD is interpolated from the aggregated data and appended to the dataset.

5. Final ND Filter Calibration Product Creation: The aggregated data, along with common metadata, is used to create an NDFilterSweetSpotDataset, which is then saved (and optionally written to disk) as a FITS file.

Functions in fluxcal.py and test_fluxcal.py are also updated to do photometry using a number of user-defined options. I have incorporated @JuergenSchreiber's feedback to assume frames are normalized for exposure time by the time those functions are called.

ajeldorado commented:
@juliamilton Are there any unit or e2e tests using corgidrp.nd_filter_calibration.create_nd_sweet_spot_dataset()? When I do a grep search for that function, it shows up only in corgidrp/nd_filter_calibration.py.

juliamilton commented:
> @juliamilton Are there any unit or e2e tests using corgidrp.nd_filter_calibration.create_nd_sweet_spot_dataset()? When I do a grep search for that function, it shows up only in corgidrp/nd_filter_calibration.py.

@ajeldorado the unit tests are in corgidrp/tests/test_nd_filter_calibration.py. They call create_nd_filter_cal(), which uses create_nd_sweet_spot_dataset()

maxwellmb commented on tests/test_nd_filter_calibration.py:
```diff
@@ -0,0 +1,396 @@
+        assert ext_hdr.get('FPAMNAME') is not None, "Missing FPAMNAME keyword."
+        assert ext_hdr.get('FPAM_H') is not None, "Missing FPAM_H keyword."
+        assert ext_hdr.get('FPAM_V') is not None, "Missing FPAM_V keyword."
+        assert ext_hdr.get('CFAMNAME') is not None, "Missing CFAMNAME keyword."
```
Since the plan for now is to propagate L1 headers this test might not be necessary. I guess it doesn't hurt though.

maxwellmb commented on tests/test_nd_filter_calibration.py:
```diff
@@ -0,0 +1,396 @@
+        dim_dataset_cached, bright_dataset_cached, output_dir,
+        FILESAVE, OD_RASTER_THRESHOLD, clean_entry,
+        transformation_matrix_file, PHOT_METHOD, FLUX_OR_IRR, PHOT_ARGS
+    )
```
I think we need to adjust this test to use NDFilterSweetSpotDataset.save() instead since that's what we'll use in operations.

maxwellmb commented on tests/test_nd_filter_calibration.py:
```diff
@@ -0,0 +1,396 @@
+    )
+    for target, data in results["flux_results"].items():
+        avg_od = data["average_od"]
+        assert abs(avg_od - INPUT_OD) < 0.3, (
```
where does the 0.3 come from? 

With the current settings don't apertures of 7 and 10 overlap with the background subtraction region? 

I'll be honest, I don't fully understand what this one is testing for? That we get the same answer regardless of the aperture? I'd imagine the aperture size will affect the result and we'll have to have some kind of adjustment applied via the frac_enc_energy. Is the idea here that the radius is big enough that the answer shouldn't dependent too much on the input? We might want to hold off on this test until we have errors propagating - you'd imagine we want the answers to agree within errors.

juliamilton replied:
Yes, I wanted to make sure that the answer isn't too dependent on radius size - meaning that either we're not catching enough encircled energy or, if it's a big radius, that the background subtraction is doing something weird. Can keep the test but make a to-do to update the actual value when errors are propagating?

maxwellmb commented on tests/test_nd_filter_calibration.py:
```diff
@@ -0,0 +1,396 @@
+    )
+    for target, data in results['flux_results'].items():
+        od_values = data['od_values']
+        allowed_scatter = 0.05
```
Where did this value come from?

maxwellmb replied:
This again seems like we'd want to validate that things are the some within errors.

juliamilton replied:
The 0.05 value is a placeholder based on a reasonable expected variation in ND filter OD. I don’t yet know what we want to define as an acceptable standard deviation in OD across all the different rasters, but we likely need a threshold that catches excessive variation. If OD fluctuates too much across all the different sampling locations, interpolating the OD for a new point may not be reliable, and we should consider re-sampling. This test is intended to catch those cases.

maxwellmb replied:
Ok, I think it's ok to leave these placeholders for now. I think in that case it'd be great to put in a comment just to mention that it's semi-arbitrary. 

But what exactly is this testing? Is it that the algorithm doesn't introduce variations above 0.05 when we simulate things at different positions? I guess there's a tradeoff between what the SNR is of the simulated data vs. algorithmic error. That's partially why I mentioned that we might want to do this test based on the errors introduced. But if we're trying to test the algorithm that's another story.

juliamilton replied:
I wrote this test as an example of how to test and flag genuine uniformity issues in the ND filter for future real data, not to measure algorithmic accuracy per se. But it does test the algorithm in the way you described (although I think that's not very useful because the simulated data is pretty deterministic). I think it's something important to test for and flag, but I can also move this functionality out of the unit tests and actually flag something in the calibration product via a header or something. 

Also if any of this is quicker to just discuss on a call feel free to do so, since there are already so many comments on this PR.

juliamilton replied:
I'm eventually going to move this into the actual calibration product generation code to flag or raise a warning if the local ODs vary too widely, but going to just mark it and open an issue as "to do for future release" for now since it's not critical to functionality.

maxwellmb commented on tests/test_nd_filter_calibration.py:
```diff
@@ -0,0 +1,396 @@
+        err_bg = np.std(results_bg['flux_results'][target]['od_values'])
+        print(f"Target {target}: OD no-bg = {od_no}, error no-bg = {err_no}; OD bg = {od_bg}, error bg = {err_bg}")
+        # The overall OD should be similar, within a small tolerance.
+        assert abs(od_no - od_bg) < 0.1, f"OD should not differ drastically between modes for {target}"
```
Similar to above: I think we want to compare against the errors in these measurements unless the 0.1 is motivated by something I've missed.

juliamilton replied:
Ok! I agree that comparing the OD difference against measurement errors is better here. The 0.1 OD threshold was a rough placeholder.

juliamilton replied:
Going to make an issue to update this when we have better errors to measure against.

maxwellmb commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,522 @@
+        phot_kwargs = {}
+
+    # 1. Average calibration factor from dim stars
+    cal_factor = compute_avg_calibration_factor(dim_stars_dataset,
```
Per the discussion in the telecon today: another option is to include both dim and bright stars in a single input dataset, but then distinguish between them based on the presence of an ND filter.

juliamilton replied:
Using this approach now, with options to either split the dataset or use a FluxcalFactor cal product

maxwellmb reviewed (changes requested):
Hi @juliamilton, 
Thanks for all the work you put into responding to my original set of comments. With the more fleshed out version here I have more comments. Please let me know if you have any questions! 
Cheers, 
Max

maxwellmb commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,522 @@
+        if not files:
+            continue
+        print(f"Processing bright target files: {target}")
+        star_data = process_bright_target(target, files, cal_factor,
```
In the end, it might held with some of the tests to include errors (as you'll see in my comments there). Open to discussion

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,522 @@
+        final_sweet_spot_data = np.vstack([final_sweet_spot_data, new_row])
+
+    # Build the NDFilterSweetSpotDataset
+    pri_hdr = fits.Header()
```
Ok - do you want that (carrying over all headers) to be the practice going forward? @nasavbailey has input that some of the headers that get automatically passed through to the calibration products are not relevant to those products, and so suggested that we only include relevant headers.

maxwellmb replied:
I'm making this an issue that we can deal with later.

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,522 @@
+    """
+    Create an NDFilterSweetSpotDataset FITS file with the Nx3 sweet-spot array.
+    Optionally append a row from a clean frame if both clean_frame_entry and
+    transformation_matrix are provided.
```
The clean frame refers to the image where we actually apply the ND filter calibration to determine the OD at that specific focal plane location (FDD req. 1090877). Per earlier discussion, the plan was to:

1. Develop the 'sweet spot dataset' using bright known stars.
2. Use this dataset and to interpolate the OD for a given image (not one of the calibrator stars) which may fall on a slightly different ND filter location than the known star dataset samples.
3. Append the EXCAM location and interpolated OD to the existing sweet spot dataset for future use.

maxwellmb replied:
My apologies if I've said something different in the past, but I don't think appending interpolated data to existing calibrations buys us anything. Adding interpolated data isn't actually new data and bookkeeping what's happened in the past for such a case sounds like a headache. 

I think likely what we might want is a step function that takes in your calibration object, and given another image can calculate the interpolated OD and add it to that image's header (or something like that). I'd suggest we push that implementation to the next PR.

juliamilton replied:
Ok. @semaphoreP and @nasavbailey initially suggested this approach, so I'll check with them to see if there's a specific reason for keeping the interpolated data in the calibration product. I agree that it doesn’t add new information, so it’s unclear to me whether storing it there has a real benefit versus just keeping it in the clean frame image file. I can implement either approach as long as there’s consensus.

nasavbailey replied:
Good points - I'm fine not saving the interpolated product.

semaphoreP replied:
Having a step function that does the interpolation sounds good to me.

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,522 @@
+    # Metadata checks
+    if (hdr.get('FPAMNAME') != ref_fpam_name or 
+        hdr.get('FPAM_H') != ref_fpam_h or 
+        hdr.get('FPAM_V') != ref_fpam_v or
```
enforced it to be within the PAM non-repeatability tolerance

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -0,0 +1,522 @@
+        ref_fpam_h (float): The reference FPAM horizontal position.
+        ref_fpam_v (float): The reference FPAM vertical position.
+        ref_cfam_name (str): The reference CFAM name for validation.
+        expected_flux (float): The expected flux value for computing OD.
```
updated

maxwellmb commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -369,87 +314,130 @@ def create_nd_sweet_spot_dataset(
+    return ndsweetspot_dataset
+
+
+def calculate_od_at_new_location(clean_frame_entry, transformation_matrix_file, 
```
Ok, I've decided to let this through and we can adjust it after Sergi's is merged. I'll make an issue.

maxwellmb reviewed (changes requested):
Ok, just one last major comment here and the rest I'm leaving as issues

juliamilton commented on corgidrp/nd_filter_calibration.py:
```diff
@@ -369,87 +314,130 @@ def create_nd_sweet_spot_dataset(
-                         output_path,
-                         file_save,
+def create_nd_filter_cal(stars_dataset,
                          od_raster_threshold,
```
Ok, I set it as a keyword argument (not keyword-only) with a default value for now. If you want them to be keyword-only, I'll need to change them. A to-do item is to determine the appropriate OD variance threshold value.

juliamilton commented:
@maxwellmb Made the header and keyword changes, resolved merge conflicts, and will leave the transformation matrix item as a to-do until Sergi's changes are merged
